protocol-test-util/src/main/java/io/camunda/zeebe/test/broker/protocol/record/ProtocolFactory.java

```diff
@@ -0,0 +1,274 @@
+ * is going to be an instance of {@link io.camunda.zeebe.protocol.record.value.JobRecordValue}, and
+ * the intent is going to be an instance of {@link
+ * io.camunda.zeebe.protocol.record.intent.JobIntent}. This is done to enable using this factory
+ * with tests that expect to serialize and deserialize the records.
+ */
+@SuppressWarnings("java:S1452")
+public final class ProtocolFactory {
+  private final CustomRandomizerRegistry randomizerRegistry;
+  private final EasyRandomParameters parameters;
+  private final EasyRandom random;
+
+  /**
+   * Every call to this constructor will always return a factory which will produce different
+   * records. If you wish to reproduce past results, consider using {@link
+   * ProtocolFactory#ProtocolFactory(long)}.
+   */
+  public ProtocolFactory() {
+    this(ThreadLocalRandom.current().nextLong());
+  }
+
+  /**
+   * Returns a factory with the given seed for the initial randomization. This enables you to
+   * reproduce past results if you know the seed. Note that you will need to also effectively
+   * reproduce the same steps as in the past; the seed by itself is not enough. So if you have a
+   * seed, and want to reproduce the 10th record generated, you will need to generate 10 records
+   * again.
+   *
+   * @param seed the seed to use
+   */
+  public ProtocolFactory(final long seed) {
+    randomizerRegistry = new CustomRandomizerRegistry();
+    parameters = getDefaultParameters().seed(seed);
+    random = new EasyRandom(parameters);
+    registerRandomizers();
+  }
+
+  /** @return a random record with a random value type */
+  public Record<RecordValue> generateRecord() {
+    return generateRecord(UnaryOperator.identity());
+  }
+
+  /**
+   * Generates a record with the given modifier applied to its builder as the last step in the
+   * generation process.
+   *
+   * @param modifier applied to the builder after all the properties have been filled as the last
+   *     step; cannot be null
+   * @return a randomly generated record
+   * @throws NullPointerException if {@code modifier} is null
+   */
+  public Record<RecordValue> generateRecord(final UnaryOperator<Builder<RecordValue>> modifier) {
+    final var valueType = random.nextObject(ValueType.class);
+    return generateRecord(valueType, modifier);
+  }
+
+  /** @return a stream of random records */
+  public Stream<Record<RecordValue>> generateRecords() {
+    return generateRecords(UnaryOperator.identity());
+  }
+
+  /**
+   * Generates a modifiable stream of records.
+   *
+   * @param modifier applied to the builder after all the properties have been filled as the last
+   *     step; cannot be null
+   * @return a stream of random records
+   * @throws NullPointerException if modifier is null
+   */
+  public Stream<Record<RecordValue>> generateRecords(
+      final UnaryOperator<Builder<RecordValue>> modifier) {
+    return Stream.generate(() -> generateRecord(modifier));
+  }
```

Comment:
🔧 To improve the readability of the class, please move these methods above the `generateRecord` methods. A similar improvement could be done by moving `generateForAllValueTypes` and related methods up. Generally, I'd expect the methods that generate streams first, followed by the methods that generate individual records, followed by the rest. This would mean that the most used methods are in the top, but also that the code reads from top to bottom (with this I mean that the generateRecords, usage the generateRecord method as part of its implementation; and that to me it feels natural to scroll down to find its implementation).